### PR TITLE
Backend connection queue

### PR DESCRIPTION
**Pull request by walid-git**
This patch allows a task to be queued when a backend reaches its max_connections. The task will queue on the backend and wait for a connection to become available, rather than immediately failing. This capability is off by default and must be enabled with new parameters.

The following parameters have been added:
`backend_wait_timeout`: the amount of time a task will wait (default 0.0). 
`backend_wait_limit`: the maximum number of tasks that can wait (default 0).

The two parameters can also be overriden for individual backends in vcl:
```
    backend foo {
        .host = "bar.com";
        .wait_timeout = 3s;
        .wait_limit = 10;
    }
```

Authored by: @drodden (with minor rearrangements)

### CODE REVIEW

**Comment by walid-git** [comment]
With this change, it seems that the default value for `thread_pool_stack` on 32 bit systems is no more sufficient (reason why `e00029.vtc` is failing on ubuntu_bionic).

**Comment by bsdphk** [comment]
As an initial matter I would prefer ``queue_length`` and ``queue_timeout`` which I think are almost self-explanatory.

I'm not sure I think it is a good idea however, but default-off mitigates that.

**Comment by Dridi** [comment]
FWIW this is a partial implementation of what we agreed on for VIP 31.

https://github.com/varnishcache/varnish-cache/wiki/VIP31%3A-Backend-connection-queue

There are two loose points not covered, disembarking and health status. Disembarking fetch tasks is a large project, one we can disregard or move to its own VIP. The saturation of both `.max_connections` and `.wait_limit`/`backend_wait_limit` is probably reasonable to address before closing VIP 31.

File: bin/varnishd/cache/cache.h
Comment (nigoroll) [inline bin/varnishd/cache/cache.h:448]: A `pthread_cond_t` is pretty heavy, can't we re-use one we already have, like `(struct boc).cond` or `(struct worker).cond`?

FTR, I have not checked the details and in particular I have not looked at the impact on code clarity, but we should we sure the footprint is worth it.

File: bin/varnishd/cache/cache_backend.c
Comment (nigoroll) [inline bin/varnishd/cache/cache_backend.c:146]: um no to `lock_it`.
Can we please have a `_locked` function, or just do the `VTAILQ_REMOVE()` wherever we want the locked variant?

File: bin/varnishd/cache/cache.h
Comment (nigoroll) [reply bin/varnishd/cache/cache.h:448]: OR, even better, couldn't the two `cw_` members live on the stack, because the wait only exists within `vbe_dir_getfd()` ?

File: include/vrt.h
Comment (nigoroll) [inline include/vrt.h:61]: The commit does not actually bump the API version, but adds VRT changes.

**Review by nigoroll — CHANGES_REQUESTED** [review]
In general, I am 👍🏽

File: include/vrt.h
Comment (Dridi) [reply include/vrt.h:61]: Because we normally bump as part of the release process.

Comment (nigoroll) [reply include/vrt.h:61]: Exactly. I was referring to the commit message [Add vrt ABI bump](https://github.com/varnishcache/varnish-cache/pull/4030/commits/88f83c44a4e5af6d790bddf0feabf15efd9ad7ef)

Comment (Dridi) [reply include/vrt.h:61]: Oh, I believe it is because a bump actually happened in the downstream commit that was cherry-picked here.

@walid-git can you please change the commit message to something like this instead?

> vrt: Register VBE connection queue changes

File: bin/varnishd/cache/cache_backend.c
Comment (drodden) [reply bin/varnishd/cache/cache_backend.c:146]: Originally the dequeue was performed at each location, doing what was needed at that spot.  But questions were raised about duplicating code, which led to this helper. I'd prefer it just go back inline where needed.

File: bin/varnishd/cache/cache.h
Comment (drodden) [reply bin/varnishd/cache/cache.h:448]: The original implementation used, I believe, `(struct boc).cond`.  But there were concerns raised about hijacking that `cond` for something like this. To alleviate those concerns it was given its own `cond`, which also helps code clarity (IMO).

Comment (drodden) [reply bin/varnishd/cache/cache.h:448]: > couldn't the two `cw_` members be packed into a struct which lives on the stack

I'm not sure I understand.  You want to put a list shared across threads on the stack?

Comment (nigoroll) [reply bin/varnishd/cache/cache.h:448]: Basically yes.
In terms of [queue(3)](https://man.netbsd.org/queue.3) which we use prefixed with `V`, there is a list head and entries linked with (double) pointers. As long as we make sure that we remove it from the list before the stack frame becomes invalid, we can sure put an entry on the stack.

Comment (drodden) [reply bin/varnishd/cache/cache.h:448]: I think it is better where it is.

Comment (Dridi) [reply bin/varnishd/cache/cache.h:448]: If we end up moving it to the stack, I'd prefer seeing it in a followup commit to keep this port aligned with the original change.

No opinion otherwise.

Comment (nigoroll) [reply bin/varnishd/cache/cache.h:448]: @drodden "I think it is better where it is" is not quite a convincing argument. My arguments are:

* The additional memory requirement for _every_ `struct busyobj` is relevant

```
(gdb) p sizeof(struct busyobj)
$1 = 440
(gdb) p sizeof(pthread_cond_t)
$2 = 48
```

* The cv is only needed in _that_ particular function, which is only called when queuing, so adding overhead to the general case (memory, initialization, destruction) is not justified if there is a sensible alternative option.

Re Dridi: sure, why not

Comment (drodden) [reply bin/varnishd/cache/cache.h:448]: @nigoroll, no it was not a very good response and I know better than to post anything before fully waking up and having my morning tea.

If you'd like to move it, I'm with Dridi in doing as a follow up.

Comment (walid-git) [reply bin/varnishd/cache/cache.h]: Moved the `cw_*` members to the stack

File: bin/varnishd/cache/cache_backend.c
Comment (walid-git) [reply bin/varnishd/cache/cache_backend.c]: Removed `vbe_connwait_dequeue` and replaced it with inlined calls

File: include/vrt.h
Comment (walid-git) [reply include/vrt.h:61]: Done

**Comment by nigoroll** [comment]
Thank you, this looks good to me overall.

I would still have some suggestions, but would also be OK to polish after merge, if you agree:

* `struct backend_cw` should have magic value which we check for every access. While I did suggest to move it to the stack, I am well aware of the risk of smashing foreign stacks, and magic checks can help raise a flag if that happens.
* the magic would be initialized with `INIT_OBJ()`, which would also initialize the list pointers for clarity. The code is still correct, IIUC, even with uninitialized list pointers, but I think it helps debugging a lot if they are zeroed.
* `cw_state` should be moved into `struct backend_cw`. This allows for additional assertions (e.g. in `vbe_connwait_signal_locked`, we can assert that the state is `CW_QUEUED`.
* when we dequeue, we should change the state to a new state, e.g. `CW_DEQUEUED`.
* We should add a `backend_cw_fini` function which asserts that the state is `!= CW_QUEUED` before destroying the condition variable.
* Regarding the inlining of the dequeue, I think I would actually prefer a `vbe_connwait_dequeue_locked` for clarity, because that would now also change the state.

**Review by nigoroll — APPROVED** [review]
see top level comment

File: bin/varnishd/cache/cache_backend.c
Comment (nigoroll) [reply bin/varnishd/cache/cache_backend.c]: see top level comment

File: bin/varnishd/cache/cache_backend.c
Comment (Dridi) [inline bin/varnishd/cache/cache_backend.c:147]: Why not `vbe_conwait_fini()`?

File: bin/varnishd/cache/cache_backend.c
Comment (Dridi) [inline bin/varnishd/cache/cache_backend.c:150]: Use the `assert()` macro if you pass an actual condition.

File: bin/varnishd/cache/cache_backend.c
Comment (Dridi) [inline bin/varnishd/cache/cache_backend.c:152]: Please `ZERO_OBJ()` for good measure.

**Review by Dridi — CHANGES_REQUESTED** [review]

File: bin/varnishd/cache/cache_backend.c
Comment (Dridi) [reply bin/varnishd/cache/cache_backend.c]: It will work in this function that takes a `struct backend_cw *cw` ;-)

File: bin/varnishd/cache/cache_backend.h
Comment (Dridi) [inline bin/varnishd/cache/cache_backend.h]: I think you should be able to get away with just an opaque `struct connwait;` here and keep the actual type definitions contained in `cache_backend.c`.

Please note how I'm implicitly suggesting to rename the struct :grin:

**Review by Dridi — CHANGES_REQUESTED** [review]

File: bin/varnishd/cache/cache_backend.h
Comment (walid-git) [reply bin/varnishd/cache/cache_backend.h]: Moved the struct to `cache_backend.c` and renamed as suggested.

File: bin/varnishd/cache/cache_backend.c
Comment (walid-git) [reply bin/varnishd/cache/cache_backend.c]: Changed to `vbe_connwait_fini`.

File: bin/varnishd/cache/cache_backend.c
Comment (walid-git) [reply bin/varnishd/cache/cache_backend.c]: Fixed.

**Comment by walid-git** [comment]
Rebased and addressed all review comments. Ready for a (hopefully) last review.

File: bin/varnishd/cache/cache_backend.c
Comment (Dridi) [inline bin/varnishd/cache/cache_backend.c:70]: Didn't we conclude that we could use `bo->wrk->cond` for the connection queue?

Comment (walid-git) [reply bin/varnishd/cache/cache_backend.c:70]: @Dridi that's not what I understood from the discussion [here](https://github.com/varnishcache/varnish-cache/pull/4030#discussion_r1430199392)
See Daryl's comment:

> The original implementation used, I believe, (struct boc).cond. But there were concerns raised about hijacking that cond for something like this. To alleviate those concerns it was given its own cond, which also helps code clarity (IMO).

Comment (Dridi) [reply bin/varnishd/cache/cache_backend.c:70]: Right, but this was about the boc cond, not the worker cond, right? It looks like we never discussed the latter.

Comment (walid-git) [reply bin/varnishd/cache/cache_backend.c:70]: It might have been discussed on the original implementation, not sure, maybe @drodden can provide some input ?

Comment (drodden) [reply bin/varnishd/cache/cache_backend.c:70]: I vaguely remember something about not using the worker cond and found this comment on the original PR:

> [daghf] [Jan 16, 2023] 
> Indeed. I also like the idea of not repurposing wrk->cond, this rhymes better with the stated future task of task disembarking.

File: bin/varnishd/cache/cache_backend.c
Comment (Dridi) [inline bin/varnishd/cache/cache_backend.c]: `Lck_AssertHeld()` ?

**Review by Dridi — APPROVED** [review]

**Review by nigoroll — CHANGES_REQUESTED** [review]
I always feel bad when it looks like I was holding things up, so I would like to apologize for not having spotted some issues earlier.

File: bin/varnishd/cache/cache_backend.c
Comment (walid-git) [reply bin/varnishd/cache/cache_backend.c:312]: Good catch, will add it, thanks.

File: bin/varnishd/cache/cache_backend.c
Comment (walid-git) [reply bin/varnishd/cache/cache_backend.c:251]: Why would we allocate workspace before we are sure that we can get a backend connection ?

File: bin/varnishd/cache/cache_backend.c
Comment (walid-git) [reply bin/varnishd/cache/cache_backend.c:187]: Fixed

File: bin/varnishd/cache/cache_backend.c
Comment (walid-git) [reply bin/varnishd/cache/cache_backend.c]: Fixed

File: bin/varnishd/cache/cache_backend.c
Comment (walid-git) [reply bin/varnishd/cache/cache_backend.c:228]: I have pushed a new commit where I removed the EINTR check and added a check for `BE_BUSY(bp)` when we wake up after a timeout. Let me know if this works for you.

Comment (Dridi) [reply bin/varnishd/cache/cache_backend.c:228]: Bugwash discussion: status quo on spurious EINTR vs cond wait.

**Review by Dridi — APPROVED** [review]

File: bin/varnishd/cache/cache.h
Comment (walid-git) [reply bin/varnishd/cache/cache.h]: Fixed

File: bin/varnishd/cache/cache_busyobj.c
Comment (walid-git) [reply bin/varnishd/cache/cache_busyobj.c]: Fixed

File: bin/varnishd/cache/cache_backend.c
Comment (walid-git) [reply bin/varnishd/cache/cache_backend.c:252]: I have added an XXX comment as suggested

Comment (nigoroll) [reply bin/varnishd/cache/cache_backend.c:252]: > On the other hand, if we are effectively not connecting and workspace would have been too tight, then we fail for the wrong reason, and we don't visit vcl_backend_error at all.

I disagree on "fail for the wrong reason". This code only runs because we do want to connect and having enough workspace is a precondition for the connect to succeed. Potentially running into the connection or wait limit does not make it a "wrong reason" to fail for insufficient workspace.

**Comment by nigoroll** [comment]
bugwash:

 proposed (re)name:

global parameters:
* `backend_queue_limit`
* `backend_queue_timeout`

VCL:

```vcl
backend foo {
  .queue_limit = 42;
  .queue_timeout 1m;
}
sub vcl_backend_fetch {
  set bereq.queue_limit = 42;
  set bereq.queue_timeout = 1m;
}
```

**Comment by gquintard** [comment]
hi all! I'd really like this to get into the next release, and from what I'm reading, it's only a naming exercise from now on.

As a refresher, the current PR offers `backend_wait_timeout`/`backend_wait_limit` and the bugwash proposed `backend_queue_timeout`/`backend_queue_limit`. I feel like this isn't a big enough contention point to block a merge.

Any chance to get the original names in? I hate to bring it up and I'll probably get slapped for it but: we have customers using the feature and consistency is pretty important, I don't want to have to translate parameter names depending on the platform people are running.

**Comment by nigoroll** [comment]
bugwash approved.
please implement connection purge on health change to sick

**Comment by dridi** [comment]
Bugwash: when a probe transitions effectively to sick, the queue should be cleared to let tasks fail immediately.

Effective transitions to sick:

- probe goes sick, not overridden by CLI (`backend.set_health healthy`)
- `backend.set_health sick` (where previously healthy from probe or CLI)
- `backend.set_health probe` (with sick probe, where previously healthy from CLI)

**Comment by walid-git** [comment]
PR updated:
- [e8f35fb](https://github.com/varnishcache/varnish-cache/pull/4030/commits/e8f35fbf2af0c7bbf78c76e661136b0f75e9734c) changes wait_timeout to match with new timeout defaults.
- [871f784](https://github.com/varnishcache/varnish-cache/pull/4030/commits/871f784f174f93f6777b5e7ba66d48be239f61e4) Notify client requests in the wait queue when the backend goes sick (as agreed during bugwash)

**Comment by nigoroll** [comment]
bugwash: merge but the last commit, open new PR for the flush

**Comment by walid-git** [comment]
As per bugwash: squashed [e8f35fb](https://github.com/varnishcache/varnish-cache/pull/4030/commits/e8f35fbf2af0c7bbf78c76e661136b0f75e9734c) and removed [871f784](https://github.com/varnishcache/varnish-cache/pull/4030/commits/871f784f174f93f6777b5e7ba66d48be239f61e4) from this PR (to be submitted separately).
Will merge once CCI is done.

**Pull request merged**
